src/execution/operator/aggregate/physical_ungrouped_aggregate.cpp

```diff
@@ -98,38 +215,104 @@ class SimpleAggregateLocalState : public LocalSinkState {
 	ExpressionExecutor child_executor;
-	//! The payload chunk
+	//! The payload chunk, containing all the Vectors for the aggregates
 	DataChunk payload_chunk;
 	//! Aggregate filter data set
 	AggregateFilterDataSet filter_set;
+	//! The local sink states of the distinct aggregates hash tables
+	vector<unique_ptr<LocalSinkState>> radix_states;
+
+public:
+	void Reset() {
+		payload_chunk.Reset();
+	}
+	void InitializeDistinctAggregates(const DistinctAggregateData &data, ExecutionContext &context) {
+
+		auto &distinct_indices = data.Indices();
+		if (distinct_indices.empty()) {
+			// No distinct aggregates
+			return;
+		}
+		D_ASSERT(!data.radix_tables.empty());
+
+		idx_t aggregate_cnt = data.radix_tables.size();
+		radix_states.resize(aggregate_cnt);
+
+		for (auto &idx : distinct_indices) {
+			auto &radix_table = *data.radix_tables[idx];
+			radix_states[idx] = radix_table.GetLocalSinkState(context);
+		}
+	}
 };
 
 unique_ptr<GlobalSinkState> PhysicalUngroupedAggregate::GetGlobalSinkState(ClientContext &context) const {
-	return make_unique<SimpleAggregateGlobalState>(aggregates);
+	return make_unique<SimpleAggregateGlobalState>(Allocator::Get(context), aggregates, distinct_aggregate_data,
+	                                               context);
 }
 
 unique_ptr<LocalSinkState> PhysicalUngroupedAggregate::GetLocalSinkState(ExecutionContext &context) const {
-	return make_unique<SimpleAggregateLocalState>(Allocator::Get(context.client), aggregates, children[0]->GetTypes());
+	return make_unique<SimpleAggregateLocalState>(Allocator::Get(context.client), aggregates, children[0]->GetTypes(),
+	                                              distinct_aggregate_data, context);
+}
+
+void PhysicalUngroupedAggregate::SinkDistinct(ExecutionContext &context, GlobalSinkState &state, LocalSinkState &lstate,
+                                              DataChunk &input) const {
+	auto &sink = (SimpleAggregateLocalState &)lstate;
+	auto &global_sink = (SimpleAggregateGlobalState &)state;
+	auto &distinct_indices = distinct_aggregate_data.Indices();
+	if (distinct_indices.empty()) {
+		return;
+	}
+	for (auto &idx : distinct_indices) {
+		auto &aggregate = (BoundAggregateExpression &)*aggregates[idx];
+
+		D_ASSERT(this->distinct_aggregate_data.radix_tables[idx]);
+		auto &radix_table = *distinct_aggregate_data.radix_tables[idx];
+		auto &radix_global_sink = *global_sink.radix_states[idx];
+		auto &radix_local_sink = *sink.radix_states[idx];
+
+		if (aggregate.filter) {
+			// Apply the filter before inserting into the hashtable
+			auto &filtered_data = sink.filter_set.GetFilterData(idx);
+			idx_t count = filtered_data.ApplyFilter(input);
+			filtered_data.filtered_payload.SetCardinality(count);
+
+			radix_table.Sink(context, radix_global_sink, radix_local_sink, filtered_data.filtered_payload,
+			                 filtered_data.filtered_payload);
+		} else {
+			radix_table.Sink(context, radix_global_sink, radix_local_sink, input, input);
```

Comment:
It seems that the `ForceSingleHT` is being triggered in the radix partitioned hash table - when this should not be required as the child hash table should not actually compute a distinct aggregate. This is preventing a lot of parallelism as all the threads are sinking into a large shared HT instead of having their own thread-local HT. Could you have a look at that?

Reply:
Ah, I still have to see if that causes any bugs/issues, but that bumped performance by quite a bit
```
name    run     timing
benchmark/micro/aggregate/simple_distinct.benchmark     1       0.050440
benchmark/micro/aggregate/simple_distinct.benchmark     2       0.050602
benchmark/micro/aggregate/simple_distinct.benchmark     3       0.051067
benchmark/micro/aggregate/simple_distinct.benchmark     4       0.051073
benchmark/micro/aggregate/simple_distinct.benchmark     5       0.050849
```

Reply:
I think this should be fixed now!